<!DOCTYPE html>
<!-- saved from url=(0067)http://127.0.0.1:5000/answerHTML/LmRaC-2024_09_04-10_40_22-005.html -->
<html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <title>LmRaC-2024_09_04-10_40_22-005</title>
    <style>
        body, html {
            height: 100%;
            margin: 0px;
            padding: 10pt;
            background-color: white;
            font-family:Arial, Helvetica, sans-serif;
            line-height: 1.6;
        }
    </style>
</head>

<body>
<h1>QUESTION</h1>
<br>
What are key genes for which we'd expect to see differential expression in the chemokine signaling pathway in colorectal cancer?<br>
<br>
<h1>ANSWER</h1>
<br>
Key genes expected to show differential expression in the chemokine signaling pathway in colorectal cancer include:<br>
<br>
1. <strong>CXCL13/CXCR5</strong>: This axis is involved in the pathogenesis, development, and metastasis of colorectal cancer through mechanisms such as upregulating MMP13 and activating the PI3K/Akt pathway (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8708574/#:~:text=In%20colorectal%20or%20colon%20cancer%2C%20the%20CXCL13,clinical%20treatment%20of%20colorectal%20cancer">Gao, SH. et al., 2021</a>).<br>
<br>
2. <strong>CXCL12/CXCR4</strong>: CXCR4 is overexpressed in colorectal carcinoma tissues and is associated with disease progression and reduced survival. CXCL12 enhances chemotaxis, adhesion, proliferation, and ERK1/2 phosphorylation in colorectal cancer cells (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2528256/#:~:text=Ottaiano,progression%20and%20reduced%20survival%20%5B%2C%20%2C%20%2C">Richard, CL. et al., 2008</a>). Additionally, CXCL12 expression is often reduced in colorectal cancer due to DNA hypermethylation, which facilitates tumor cell migration (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8997717/#:~:text=One%20hypothesis%20is%20that%20before%20metastasis,contact%20is%20lost%20between%20epithelial%20cells">Goïta, AA. et al., 2022</a>).<br>
<br>
3. <strong>CXCL8/CXCR1/2</strong>: CXCL8 is markedly upregulated in colorectal carcinoma and is associated with tumor cell proliferation, migration, angiogenesis, and chemoresistance. The CXCL8-CXCR1/2 axis is crucial for tumor progression and metastasis (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6142815/#:~:text=Apart%20from%20the%20cancer%20types%20discussed,sunitinib%20treatment%20in%20a%20nude%20mouse%20model">Liu, Q. et al., 2016</a>)(<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5436513/#:~:text=Our%20bioinformatics%20analysis%20reveals,the%20CXCL8%2DCXCR1/2%20signaling%20pathways">Ha, H. et al., 2017</a>).<br>
<br>
4. <strong>CXCL2</strong>: CXCL2 is significantly overexpressed in colorectal cancer and is involved in tumor initiation, development, and angiogenesis. Knockdown of CXCL2 reduces the expression of EMT markers and MMPs, thereby inhibiting tumor proliferation and migration (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7522359/#:~:text=CXCL2%20belongs%20to%20the%20chemokine,for%20verification%20in%20future%20studies">Liu, Y. et al., 2020</a>).<br>
<br>
5. <strong>CXCL9, CXCL10, and CXCL13</strong>: These chemokines are associated with high densities of Th1/cytotoxic memory T cells and favorable prognosis in colorectal cancer. They are involved in building an efficient immune contexture at the tumor site (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3044219/#:~:text=An%20obvious%20question%20that%20has%20both,%2C%20such%20as%20CXCL5%20for%20instance%2C%20do%20not">Mlecnik, B. et al., 2011</a>)(<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9660543/#:~:text=Next%2C%20we%20stratified%20the%20TCGA%20samples%20into,gene%20sets%20%28Supplementary%20Fig.%20S4">Kocher, F. et al., 2022</a>).<br>
<br>
6. <strong>CCL19/CCR7</strong>: CCL19 inhibits tumorigenesis, metastasis, and angiogenesis in colorectal cancer and is associated with a good prognosis. It blocks angiogenesis by inhibiting the Met/ERK/Elk-1/HIF-1α/VEGF-A pathway in a CCR7-dependent manner (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8508474/#:~:text=Therefore%2C%20TME%20has%20pro%2Dtumorigenic,%2DA%20pathway%20in%20a%20CCR7%2Ddependent%20pattern">Stefani, C. et al., 2021</a>).<br>
<br>
These genes play significant roles in the chemokine signaling pathway and are critical for understanding the molecular mechanisms underlying colorectal cancer progression and potential therapeutic targets.<br>
<br>
Notable genes and their associated proteins in the chemokine signaling pathway include:<br>
<br>
1. <strong>Chemokines and Chemokine Receptors:</strong><br>
   - <strong>CXCL12</strong> and its receptors <strong>CXCR4</strong> and <strong>ACKR3</strong> (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6718456/#:~:text=A%20good%20example%20of%20the%20complex%20biology%20of,in%20several%20autoimmune%20diseases">García-Cuesta, EM. et al., 2019</a>).<br>
   - <strong>CCL2 (MCP-1)</strong> and its receptor <strong>CCR2</strong> (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9811007/#:~:text=Chemokines%20are%20a%20category%20of%20minor,of%20cytokines%20and%20adhesion%20molecules%20%28%2C">Yu, W. et al., 2022</a>).<br>
   - <strong>CCL5 (RANTES)</strong>, <strong>CCL20 (MIP-3α)</strong>, <strong>CXCL8 (IL-8)</strong>, <strong>CXCL16 (SR-PSOX)</strong>, and <strong>CX3CL1 (fractalkine)</strong> (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9811007/#:~:text=Chemokines%20are%20a%20category%20of%20minor,of%20cytokines%20and%20adhesion%20molecules%20%28%2C">Yu, W. et al., 2022</a>).<br>
   - <strong>CXCR1, CXCR2, CXCR3, CXCR4, CXCR5</strong> (CXC receptors) and <strong>CCR1, CCR2, CCR3, CCR4, CCR5, CCR6, CCR7, CCR8, CCR9</strong> (CC receptors) (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4818359/#:~:text=,">Ahmadiankia, N. et al., 2016</a>).<br>
<br>
2. <strong>Signaling Pathways and Molecules:</strong><br>
   - <strong>PI3K/AKT pathway</strong>: Genes such as <strong>PIK3R3</strong> and <strong>AKT</strong> (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3908308/#:~:text=Based%20on%20the%20existing%20information%2C%20we,mutants%20are%3B%20PIP5KIA%20and%20PIP5KIC%20genes">Sharma, M. et al., 2013</a>)(<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7226115/#:~:text=It%20is%20thought%20that%20chemokine%20responses,and%20the%20activation%20of%20survival%20mechanisms">Rizeq, B. et al., 2020</a>).<br>
   - <strong>MAPK pathway</strong>: Genes such as <strong>MAPK4, MAPK9, MAP3K9</strong>, and components like <strong>ERK1/2</strong> and <strong>p38</strong> (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3908308/#:~:text=Based%20on%20the%20existing%20information%2C%20we,mutants%20are%3B%20PIP5KIA%20and%20PIP5KIC%20genes">Sharma, M. et al., 2013</a>)(<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6063446/#:~:text=Most%20of%20the%20chemokines%20expression%20is,are%20reported%20to%20activate%20p38%20and%20JNK%2C">Verma, AK. et al., 2018</a>).<br>
   - <strong>JAK/STAT pathway</strong>: Genes encoding <strong>Janus kinases (JAKs)</strong> and <strong>signal transducers and activators of transcription (STATs)</strong> (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2193934/#:~:text=After%20binding%20to%20their%20receptors,STAT%29%20transcription%20factors">Soriano, SF. et al., 2002</a>)(<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7226115/#:~:text=It%20is%20thought%20that%20chemokine%20responses,and%20the%20activation%20of%20survival%20mechanisms">Rizeq, B. et al., 2020</a>).<br>
   - <strong>NF-κB pathway</strong>: Genes involved in the activation of <strong>NF-κB</strong> and its regulatory components (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6063446/#:~:text=Most%20of%20the%20chemokines%20expression%20is,are%20reported%20to%20activate%20p38%20and%20JNK%2C">Verma, AK. et al., 2018</a>)(<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7769120/#:~:text=We%20used%20the%20Kyoto%20Encyclopedia%20of%20Genes,process%20of%20tumor%20invasion%20and%20metastasis">Lu, C. et al., 2020</a>).<br>
   - <strong>PLC pathway</strong>: Genes encoding <strong>phospholipase C (PLC)</strong> and related molecules like <strong>PIP2, DAG, and IP3</strong> (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8910147/#:~:text=Chemokines%20are%20grouped%20into%20four%20distinct,promoting%20cell%20survival%20and%20chemotaxis">Huang, X. et al., 2022</a>).<br>
<br>
3. <strong>Other Key Genes:</strong><br>
   - <strong>ADAM10</strong>: A protease responsible for cleaving CXCL16, which is a proposed diagnostic and prognostic marker for various cancers (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6385302/#:~:text=Years%20of%20research%20have%20shown%20that%20a%20very,and%20poor%20prognosis%20of%20gastric%20cancer%E2%80%93">Mir, H. et al., 2019</a>).<br>
<br>
These genes and their associated pathways play crucial roles in mediating the effects of chemokines on cell migration, proliferation, survival, and other cellular functions, particularly in the context of inflammation, immune responses, and cancer progression.<br>
<br>
The genes that show differential expression in colorectal cancer (CRC) include:<br>
<br>
1. <strong>NAT10</strong>: Identified as a regulator of mRNA acetylation with multiple functions in cancer (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10026876/#:~:text=We%20identified%20genes%20differentially,normal%20and%20tumour%29%20of%20differential%20genes">Zhang, H. et al., 2023</a>).<br>
2. <strong>LGR5, CLDN9, MYC, MEX3A</strong>: Differentially expressed between normal and CRC tissues, as well as between high and low NAT10 expression groups (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10026876/#:~:text=Identification%20of%20NAT10%2Drelated%20ac4C%2DDEGs,normal%20and%20tumour%29%20of%20differential%20genes">Zhang, H. et al., 2023</a>).<br>
3. <strong>ARHGEF7</strong>: Overexpressed in CRC and associated with metastasis and poor survival (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7140047/#:~:text=Immunostaining%20of%20ARHGEF7%20%28beta%2DPIX%29%20is,of%20colorectal%20cancer%20cell%20lines">Kotelevets, L. et al., 2020</a>).<br>
4. <strong>APC, KRAS, PIK3CA, GPT2, MSH2, MLH1, TP53, CTNNB1, AKT1, MMP2, NRAS, CD274, STK11, CTLA4, PARK7, SNCA</strong>: Found to have somatic mutations in CRC (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10557533/#:~:text=,">Cao, Y. et al., 2023</a>).<br>
5. <strong>SFRP4, AURKA, BMP7, CENPH, PTGER4, ID4</strong>: Differentially expressed between adenomas and carcinomas (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3308003/#:~:text=mRNA%20expression%20was%20significantly,by%20Skrzypczak%20and%20collaborators">Carvalho, B. et al., 2012</a>).<br>
6. <strong>NAT2, XDH, GPX3, AOC3, AKR1C4, SPHK1, ADCY5</strong>: Metabolism-related genes significantly related to the overall survival of CRC patients (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7474523/#:~:text=A%20total%20of%20102%20differentially%20expressed,were%20high%2Drisk%20PRMGs%20%28HR%20%3E%201%2C%20p%C2%A0%3C%C2%A00.05">Miao, Y. et al., 2020</a>).<br>
7. <strong>JAG2</strong>: Increased expression in CRC cell lines compared to normal colorectal cell lines (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6558914/#:~:text=Next%2C%20the%20differential%20expression%20of%20JAG2,with%20that%20of%20the%20CCD18%2DCo%20%28Fig.%C2%A0b%2C%20c">He, W. et al., 2019</a>).<br>
8. <strong>NEUROG1, REG1B, CMIP, KPNA3</strong>: Identified as differentially expressed genes in CRC (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10431729/#:~:text=Differential%20analysis%20of%20the%20GSE68204,was%20run%20in%20triplicate%20independently">Wang, XW. et al., 2023</a>).<br>
9. <strong>APC, BAX, BCL2, DCC, KRAS, MLH1, MSH2, MSH6, SMAD2, SMAD4, TGFBR2, TP53</strong>: Frequently involved in CRC and defined in KEGG colorectal cancer MSI and CIN signaling pathways (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4033420/#:~:text=In%20order%20to%20discover%20more%20potential,dataset%20GSE35982%20for%20regression%20analysis">Wang, F. et al., 2014</a>).<br>
<br>
These genes have been identified through various studies and datasets, indicating their significant roles in the pathogenesis and progression of colorectal cancer.<br>
<br>
<h2>REFERENCES</h2>
<br>
Ahmadiankia, N; Moghaddam, HK; Mishan, MA; Bahrami, AR; Naderi-Meshkin, H; Bidkhori, HR; Moghaddam, M; Mirfeyzi, SJ. Berberine suppresses migration of MCF-7 breast cancer cells through down-regulation of chemokine receptors. Iranian journal of basic medical sciences. 2016.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4818359/">PMC4818359</a><br>
<br>
Cao, Y; Wang, J; Hou, W; Ding, Y; Zhu, Y; Zheng, J; Huang, Q; Cao, Z; Xie, R; Wei, Q; Qin, H. Colorectal cancer-associated T cell receptor repertoire abnormalities are linked to gut microbiome shifts and somatic cell mutations. Gut microbes. 2023.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10557533/">PMC10557533</a><br>
<br>
Carvalho, B; Sillars-Hardebol, AH; Postma, C; Mongera, S; Terhaar Sive Droste, J; Obulkasim, A; van de Wiel, M; van Criekinge, W; Ylstra, B; Fijneman, RJ; Meijer, GA. Colorectal adenoma to carcinoma progression is accompanied by changes in gene expression associated with ageing, chromosomal instability, and fatty acid metabolism. Cellular oncology (Dordrecht, Netherlands). 2012.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3308003/">PMC3308003</a><br>
<br>
Gao, SH; Liu, SZ; Wang, GZ; Zhou, GB. CXCL13 in Cancer and Other Diseases: Biological Functions, Clinical Significance, and Therapeutic Opportunities. Life (Basel, Switzerland). 2021.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8708574/">PMC8708574</a><br>
<br>
García-Cuesta, EM; Santiago, CA; Vallejo-Díaz, J; Juarranz, Y; Rodríguez-Frade, JM; Mellado, M. The Role of the CXCL12/CXCR4/ACKR3 Axis in Autoimmune Diseases. Frontiers in endocrinology. 2019.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6718456/">PMC6718456</a><br>
<br>
Goïta, AA; Guenot, D. Colorectal Cancer: The Contribution of CXCL12 and Its Receptors CXCR4 and CXCR7. Cancers. 2022.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8997717/">PMC8997717</a><br>
<br>
Ha, H; Debnath, B; Neamati, N. Role of the CXCL8-CXCR1/2 Axis in Cancer and Inflammatory Diseases. Theranostics. 2017.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5436513/">PMC5436513</a><br>
<br>
He, W; Tang, J; Li, W; Li, Y; Mei, Y; He, L; Zhong, K; Xu, R. Mutual regulation of JAG2 and PRAF2 promotes migration and invasion of colorectal cancer cells uncoupled from epithelial-mesenchymal transition. Cancer cell international. 2019.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6558914/">PMC6558914</a><br>
<br>
Huang, X; Hao, J; Tan, YQ; Zhu, T; Pandey, V; Lobie, PE. CXC Chemokine Signaling in Progression of Epithelial Ovarian Cancer: Theranostic Perspectives. International journal of molecular sciences. 2022.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8910147/">PMC8910147</a><br>
<br>
Kocher, F; Puccini, A; Untergasser, G; Martowicz, A; Zimmer, K; Pircher, A; Baca, Y; Xiu, J; Haybaeck, J; Tymoszuk, P; Goldberg, RM; Petrillo, A; Shields, AF; Salem, ME; Marshall, JL; Hall, M; Korn, WM; Nabhan, C; Battaglin, F; Lenz, HJ; Lou, E; Choo, SP; Toh, CK; Gasteiger, S; Pichler, R; Wolf, D; Seeber, A. Multi-omic Characterization of Pancreatic Ductal Adenocarcinoma Relates CXCR4 mRNA Expression Levels to Potential Clinical Targets. Clinical cancer research : an official journal of the American Association for Cancer Research. 2022.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9660543/">PMC9660543</a><br>
<br>
Kotelevets, L; Chastre, E. Rac1 Signaling: From Intestinal Homeostasis to Colorectal Cancer Metastasis. Cancers. 2020.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7140047/">PMC7140047</a><br>
<br>
Liu, Q; Li, A; Tian, Y; Wu, JD; Liu, Y; Li, T; Chen, Y; Han, X; Wu, K. The CXCL8-CXCR1/2 pathways in cancer. Cytokine &amp; growth factor reviews. 2016.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6142815/">PMC6142815</a><br>
<br>
Liu, Y; Li, C; Dong, L; Chen, X; Fan, R. Identification and verification of three key genes associated with survival and prognosis of COAD patients via integrated bioinformatics analysis. Bioscience reports. 2020.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7522359/">PMC7522359</a><br>
<br>
Lu, C; Zhu, J; Chen, X; Hu, Y; Xie, W; Yao, J; Huang, S. Risk Stratification in Acute Myeloid Leukemia Using CXCR Gene Signatures: A Bioinformatics Analysis. Frontiers in oncology. 2020.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7769120/">PMC7769120</a><br>
<br>
Miao, Y; Li, Q; Wang, J; Quan, W; Li, C; Yang, Y; Mi, D. Prognostic implications of metabolism-associated gene signatures in colorectal cancer. PeerJ. 2020.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7474523/">PMC7474523</a><br>
<br>
Mir, H; Kaur, G; Kapur, N; Bae, S; Lillard, JW; Singh, S. Higher CXCL16 exodomain is associated with aggressive ovarian cancer and promotes the disease by CXCR6 activation and MMP modulation. Scientific reports. 2019.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6385302/">PMC6385302</a><br>
<br>
Mlecnik, B; Bindea, G; Pagès, F; Galon, J. Tumor immunosurveillance in human cancers. Cancer metastasis reviews. 2011.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3044219/">PMC3044219</a><br>
<br>
Richard, CL; Blay, J. CXCR4 in Cancer and Its Regulation by PPARgamma. PPAR research. 2008.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2528256/">PMC2528256</a><br>
<br>
Rizeq, B; Malki, MI. The Role of CCL21/CCR7 Chemokine Axis in Breast Cancer Progression. Cancers. 2020.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7226115/">PMC7226115</a><br>
<br>
Sharma, M; Afrin, F; Tripathi, R; Gangenahalli, G. Regulated expression of CXCR4 constitutive active mutants revealed the up-modulated chemotaxis and up-regulation of genes crucial for CXCR4 mediated homing and engraftment of hematopoietic stem/progenitor cells. Journal of stem cells &amp; regenerative medicine. 2013.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3908308/">PMC3908308</a><br>
<br>
Soriano, SF; Hernanz-Falcón, P; Rodríguez-Frade, JM; De Ana, AM; Garzón, R; Carvalho-Pinto, C; Vila-Coro, AJ; Zaballos, A; Balomenos, D; Martínez-A, C; Mellado, M. Functional inactivation of CXC chemokine receptor 4-mediated responses through SOCS3 up-regulation. The Journal of experimental medicine. 2002.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2193934/">PMC2193934</a><br>
<br>
Stefani, C; Miricescu, D; Stanescu-Spinu, II; Nica, RI; Greabu, M; Totan, AR; Jinga, M. Growth Factors, PI3K/AKT/mTOR and MAPK Signaling Pathways in Colorectal Cancer Pathogenesis: Where Are We Now? International journal of molecular sciences. 2021.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8508474/">PMC8508474</a><br>
<br>
Verma, AK; Waghmare, TS; Jachak, GR; Philkhana, SC; Reddy, DS; Basu, A. Nitrosporeusine analogue ameliorates Chandipura virus induced inflammatory response in CNS via NFκb inactivation in microglia. PLoS neglected tropical diseases. 2018.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6063446/">PMC6063446</a><br>
<br>
Wang, F; Wong, SC; Chan, LW; Cho, WC; Yip, SP; Yung, BY. Multiple regression analysis of mRNA-miRNA associations in colorectal cancer pathway. BioMed research international. 2014.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4033420/">PMC4033420</a><br>
<br>
Wang, XW; Jiang, YH; Ye, W; Shao, CF; Xie, JJ; Li, X. SIRT1 promotes the progression and chemoresistance of colorectal cancer through the p53/miR-101/KPNA3 axis. Cancer biology &amp; therapy. 2023.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10431729/">PMC10431729</a><br>
<br>
Yu, W; Ilyas, I; Hu, X; Xu, S; Yu, H. Therapeutic potential of paeoniflorin in atherosclerosis: A cellular action and mechanism-based perspective. Frontiers in immunology. 2022.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9811007/">PMC9811007</a><br>
<br>
Zhang, H; Shan, W; Yang, Z; Zhang, Y; Wang, M; Gao, L; Zeng, L; Zhao, Q; Liu, J. NAT10 mediated mRNA acetylation modification patterns associated with colon cancer progression and microsatellite status. Epigenetics. 2023.  <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10026876/">PMC10026876</a><br>


<iframe id="vJMUEXtJ" frameborder="0" src="./LmRaC-2024_09_04-10_40_22-005_files/translateSandbox.html" style="width: 0px; height: 0px; display: none;"></iframe></body></html>